### ISSUE TITLE
allowed_ips

### ISSUE THREAD

**Issue by steom** [xutianyi1999/fubuki#15]
there are some requirements for using allowed_ips on windows?

adding the option:

ERROR simple_wintun::adapter - Delete wintun driver error: OS Error -536870339 (FormatMessageW() returned error 317) (os error -536870339)
node xxx tcp handler error

Caused by:
    add nat record failed

**Comment by xutianyi1999** [comment]
Can `NetNat` commands be executed in powershell?

**Comment by steom** [comment]
yes, New-NetNat and Set-NetNat can be executed in powershell.

**Comment by xutianyi1999** [comment]
Are there more error logs?

**Comment by steom** [comment]
no

**Comment by xutianyi1999** [comment]
Recompile, you may see more errors

**Comment by steom** [comment]
> Are there more error logs?

+ New-NetNat -Name fubuki-172.0.0.0/24 -InternalIPInterfaceAddressPrefi ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (MSFT_NetNat:root/StandardCimv2/MSFT_NetNat) [New-NetNat], CimException
    + FullyQualifiedErrorId : HRESULT 0x80041013,New-NetNat

[2023-06-15 20:07:16] ERROR simple_wintun::adapter - Delete wintun driver error: OS Error -536870339 (FormatMessageW() returned error 317) (os error -536870339)
node t1 tcp handler error

**Comment by xutianyi1999** [comment]
May have to enable Hyper-V
https://learn.microsoft.com/en-us/virtualization/hyper-v-on-windows/user-guide/setup-nat-network

**Comment by xutianyi1999** [comment]
Add [cross_nat](https://github.com/xutianyi1999/fubuki/blob/c2aa9620a88f775ec4afbc63044c7dd8f4caff22/cfg-example/node-conf.json#L18) option

Used to replace the current platform system NAT component. It can be enabled when the system NAT is unavailable. It only takes effect when `allowed_ips` is configured. The default is false

**Comment by steom** [comment]
finally! very good!
I will give you feedback!

**Comment by xutianyi1999** [comment]
> finally! very good! I will give you feedback!

[precompiled test version](https://github.com/xutianyi1999/fubuki-test/releases/tag/v0.1.86)

**Comment by steom** [comment]
subnets ping works successfully
but 
ERROR fubukil::node::cross_nat - tcp_inbound_handler error: connect to 192.168.0.3:445 error

Caused by:
     Unable to establish connection. Incorrect response from the connected party after the time interval or no response from the connected host.  (os error 10060)

**Comment by xutianyi1999** [comment]
Not sure if you have config it correctly
when Node A has route 192.168.0.0/24, add allowed_ips config
```json
"allowed_ips": [
        "192.168.0.0/24"
]
```

when Node B wants to access 192.168.0.3, add ips config 
```json
 "ips": {
        "Node A IP": [
          "192.168.0.0/24"
        ]
}
```

**Comment by steom** [comment]
yes sure...
without this conf Node B, will not ping 192.168.0.3... 

udp traffic is ok
tcp tcp_inbound_handler error

**Comment by xutianyi1999** [comment]
Reply of this ping is fake
Can TCP be established with 192.168.0.3:445 on Node A?

**Comment by steom** [comment]
i've checked with an udp payload test ok

i've checked with an tcp payload test also...

tcp ERROR fubukil::node::cross_nat - tcp_inbound_handler error: connect to 192.168.0.3:445 error on node A

**Comment by xutianyi1999** [comment]
Only using `telnet 192.168.0.3 445` on Node A is also established?

**Comment by steom** [comment]
yes

new-Object System.Net.Sockets.TcpClient("192.168.0.3", 445)


Client              : System.Net.Sockets.Socket
Available           : 0
Connected           : True
ExclusiveAddressUse : False
ReceiveBufferSize   : 65536
SendBufferSize      : 65536
ReceiveTimeout      : 0
SendTimeout         : 0
LingerState         : System.Net.Sockets.LingerOption
NoDelay             : False

**Comment by xutianyi1999** [comment]
remove the socket network interface binding in cross_nat
https://github.com/xutianyi1999/fubuki-test/releases/tag/v0.1.87

**Comment by steom** [comment]
fine! now works ok

**Comment by xutianyi1999** [comment]
cross-nat merged into v0.5.9
